presto-main/src/main/java/com/facebook/presto/transaction/TransactionManager.java

```diff
@@ -443,14 +451,14 @@ public synchronized void checkConnectorWrite(ConnectorId connectorId)
 
         public synchronized CompletableFuture<?> asyncCommit()
         {
-            if (!completedSuccessfully.compareAndSet(null, true)) {
-                if (completedSuccessfully.get()) {
-                    // Already done
-                    return completedFuture(null);
-                }
-                // Transaction already aborted
+            if (state == State.COMMITTED) {
+                return completedFuture(null);
+            }
+            if (state == State.ABORTED) {
                 return failedFuture(new PrestoException(TRANSACTION_ALREADY_ABORTED, "Current transaction has already been aborted"));
             }
+            verify(state == State.ACTIVE);
+            state = State.COMMITTED;
```

Comment:
At the least this should have a comment about the fact this could still fail and be aborted. Or maybe change the name to something like `CLOSED_VIA_COMMIT`